File: java/ods/YFastTrie.java
package ods;

import java.util.Comparator;
import java.util.Iterator;
import java.util.Random;

public class YFastTrie<T> implements SSet<T> {
	protected static class Pair<T> {
		int x;
		STreap<T> t;
		public Pair(int x) { this.x = x; }
		public Pair(int x, STreap<T> t) { this.x = x; this.t = t; }
	}
	protected static class STreap<T> extends Treap<T> {
		/**
		 * Split this treap into a new treap containing all elements less
		 * than or equal to x and leave all elements greater than x in the 
		 * current treap -- assumes x is contained in the current treap
		 * @param x
		 * @return
		 */
		public STreap<T> split(T x) {
			Node<T> u = findLast(x);
			Node<T> s = newNode();
			if (u.right == null) {
				u.right = s;
			} else {
				u = u.right;
				while (u.left != null)
					u = u.left;
				u.left = s;
			}
			s.parent = u;
			s.p = Integer.MIN_VALUE;
			bubbleUp(s);
			this.r = s.right;
			if (this.r != null) this.r.parent = nil;
			STreap<T> ret = new STreap<T>();
			ret.r = s.left;
			if (ret.r != null) ret.r.parent = nil;
			return ret;
		}
		/**
		 * Absorb the elements of treap t, which should all be smaller than
		 * all the elements in this
		 * @param t
		 * @return
		 */
		public void absorb(STreap<T> t) {
			Node<T> s = newNode();
			s.right = this.r;
			if (this.r != null) this.r.parent = s;
			s.left = t.r;
			if (t.r != null) t.r.parent = s;
			this.r = s;
			t.r = null;
			trickleDown(s);
			splice(s);
		}
		/**
		 * Override this to throw an exception, since the value of n is not
		 * maintained on these treaps
		 */
		public int size() {
			throw new UnsupportedOperationException();
		}
		public String toString() {
			return Utils.collectionToString(this);
		}
	}
	protected static class Node<T> extends XFastTrie.Nöde<Node<T>,Pair<T>> {};
	protected static final int w = XFastTrie.w;
	protected XFastTrie<Node<T>,Pair<T>> xft;
	protected Integerizer<T> it;
	Random rand;
	int n;
	
	public YFastTrie(Integerizer<T> itx) {
		this.it = itx;
		Integerizer<Pair<T>> it2 = new Integerizer<Pair<T>>() {
			public int intValue(Pair<T> p) {
				return p.x;
			}
		};
		xft = new XFastTrie<Node<T>,Pair<T>>(new Node<T>(), it2);
		xft.add(new Pair<T>(0xffffffff, new STreap<T>()));
		rand = new Random(0);   // FIXME - for debugging only
		n = 0;
	}
	
	public boolean add(T x) {
		int ix = it.intValue(x);
		STreap<T> t = xft.find(new Pair<T>(ix)).t;
		if (t.add(x)) {
			n++;
			if (rand.nextInt(w) == 0) {
				STreap<T> t1 = t.split(x);
				xft.add(new Pair<T>(ix, t1));
			}
			return true;
		} 
		return false;
	}

	public void clear() {
		xft.clear();
		xft.add(new Pair<T>(0xffffffff, new STreap<T>()));
		n = 0;
	}

	public Comparator<? super T> comparator() {
		return new Comparator<T>() {
			public int compare(T a, T b) {
				return it.intValue(a) - it.intValue(b);
			}
		};
	}

	public T find(T x) {
		return xft.find(new Pair<T>(it.intValue(x))).t.find(x);
	}

	public T findGE(T x) {
		return xft.find(new Pair<T>(it.intValue(x))).t.find(x);
	}

	public T findLT(T x) {
		int ix = it.intValue(x);
		Node<T> u = xft.findNode(ix);
		T y = u.x.t.findLT(x);
		if (y == null && u.child[0] != xft.dummy)
			y = u.child[0].x.t.findLT(x);
		return y;
	}

	public Iterator<T> iterator(T x) {
		Iterator<Pair<T>> it1 = xft.iterator(new Pair<T>(it.intValue(x)));
		Pair<T> p = it1.next();
		return new MIterator(it1, p.t.iterator(x));
	}

	public boolean remove(T x) {
		int ix = it.intValue(x);
		Node<T> u = xft.findNode(ix);
		boolean ret = u.x.t.remove(x);
		if (ret) n--;
		if (u.x.x == ix && ix != 0xffffffff) {
			STreap<T> t2 = u.child[1].x.t;
			t2.absorb(u.x.t);
			xft.remove(u.x);
		}
		return ret;
	}

	@Override
	public int size() {
		return n;
	}

	protected class MIterator implements Iterator<T> {
		boolean over;
		Iterator<Pair<T>> it1;
		Iterator<T> it2;
		public MIterator(Iterator<Pair<T>> it1, Iterator<T> it2) {
			this.it1 = it1;
			this.it2 = it2;
		}
		public MIterator() {
			it1 = xft.iterator();
			if (it1.hasNext()) {
				it2 = it1.next().t.iterator();
			} 
		}
		public boolean hasNext() {
			return it2 != null && it2.hasNext();
		}
		public T next() {
			T x = it2.next();
			while (it2 != null && !it2.hasNext()) {
				if (it1.hasNext()) {
					it2 = it1.next().t.iterator();
				} else {
					it2 = null;
				}
			}
			return x;
		}
		public void remove() {
			throw new UnsupportedOperationException();
		}
	}

	public Iterator<T> iterator() {
		return new MIterator();
	}

	public String toString() {
		return Utils.collectionToString(this);
	}

	/**
	 * @param args
	 */
	public static void main(String[] args) {
		class I implements Integerizer<Integer> { 
			public int intValue(Integer i) { return i; }
		}
		int n = 50;
		YFastTrie<Integer> t = new YFastTrie<Integer>(new I());
		Random rand = new Random(0);
		System.out.println("Adding: ");
		for (int i = 0; i < n; i++) {
			int x = rand.nextInt(100*n);
			System.out.print(x + ((i < n - 1) ? "," : ""));
			t.add(x);
			// t.checkIt();
		}
		System.out.println();
		for (Pair<Integer> p : t.xft) {
			System.out.print(p.x + ",");
		}
		System.out.println();
		for (Pair<Integer> p : t.xft) {
			System.out.print(p.t + ",");
		}

		System.out.println(t);
		System.out.print("Searching: ");
		for (int i = 0; i < n; i++) {
			int x = rand.nextInt(100*n);
			System.out.print(x + "=>" + t.find(x) + ",");
		}
		System.out.println();
		System.out.println(t);
		System.out.print("Removing: ");
		for (int i = 0; i < n/2; i++) {
			Integer x = t.find(rand.nextInt(100*n));
			if (x != null) {
				System.out.print(x + ((i < n/2-1) ? "," : ""));
				System.out.flush();
				t.remove(x);
			}
			// t.checkIt();
		}
		System.out.println();
		System.out.println("Size = " + t.size());
		System.out.println(t);
		System.out.print("Searching: ");
		for (int i = 0; i < n; i++) {
			int x = rand.nextInt(100*n);
			System.out.print(x + "=>" + t.find(x) + ",");
		}
		System.out.println();
		System.out.println("done");
	}


}

File: python/ods/yfasttrie.py
"""An implementation of Willard's Y-Fast tries

This structure is able to store w-bit integers with O(log w) time searches,
additions, and removals.

D. E. Willard. Log-logarithmic worst-case range queries are possible in 
  space Theta(n). Information Processing Letters, 17, 81-84. 1984.
"""
import random

from base import BaseSet
from treap import Treap
from xfasttrie import XFastTrie
from utils import w

class STreap(Treap):
    """A Treap that implements the split/absorb functionality"""
    
    def split(self, x):
        """Remove all values <= x and return a STreap containing these values"""
        u = self._find_last(x)
        s = self._new_node(None)
        if u.right is self.nil:
            u.right = s
        else:
            u = u.right
            while u.left is not self.nil:
                u = u.left
            u.left = s
        s.parent = u
        s.p = -1
        self.bubble_up(s)
        self.r = s.right
        if self.r is not self.nil: self.r.parent = self.nil
        ret = STreap()
        ret.r = s.left
        if ret.r is not ret.nil: ret.r.parent = ret.nil
        return ret
    
    def absorb(self, t):
        """Absorb the treap t (which must only contain smaller values)"""
        s = self._new_node(None)
        s.right = self.r
        if self.r is not self.nil: self.r.parent = s
        s.left = t.r
        if t.r is not t.nil: t.r.parent = s
        self.r = s
        t.r = t.nil
        self.trickle_down(s)
        self.splice(s)
    
    def size(self):
        """Raise an error because our implementation is only half-assed"""
        raise AttributeError(self.__class__.__name__ \
                             + "does not correctly maintain size()")

class Pair(tuple):
    def __new__(cls, x, y=None):
        return super(Pair, cls).__new__(cls, (x, y))
    
    @property
    def t(self):
        return self[1]
    
    @property
    def x(self):
        return self[0]
    
    def __int__(self):
        return int(self[0])
    
class YFastTrie(BaseSet):
    def __init__(self):
        super(YFastTrie, self).__init__()
        self._initialize()
        
    def _initialize(self):
        self.xft = XFastTrie()
        self.xft.add(Pair((1<<w)-1, STreap()))
        self.n = 0
    
    def add(self, x):
        ix = int(x)
        t = self.xft.find(Pair(ix))[1]
        if t.add(x):
            self.n += 1
            if random.randrange(w) == 0:
                t1 = t.split(x)
                self.xft.add(Pair(ix, t1))
            return True
        return False
    
    def find(self, x):
        return self.xft.find(Pair(int(x)))[1].find(x)
    
    def remove(self, x):
        ix = int(x)
        u = self.xft.find_node(ix)
        ret = u.x[1].remove(x)
        if ret: self.n -= 1
        if u.x[0] == ix and ix != (1<<w) - 1:
            t2 = u.next.x[1]
            t2.absorb(u.x[1])
            self.xft.remove(u.x)
        return ret
        
    def __iter__(self):
        # self.xft is a bunch of pairs
        for p in self.xft:
            # the one'th element of each pair is an STreap
            for x in p[1]:
                yield x
    
    def clear(self):
        self._initialize()
        
    